firebase-firestore/src/main/java/com/google/firebase/firestore/util/TaskQueue.java

```diff
@@ -0,0 +1,120 @@
+ * `awaitResults()`. It is an error to enqueue tasks after invoking `awaitResults()`.
+ *
+ * <p>This class is not thread-safe. All public methods must be called from the same thread.
+ */
+public class TaskQueue<T> {
+  private BlockingQueue<TaskResult<T>> taskResults = new PriorityBlockingQueue<>();
+  private Semaphore completedTasks = new Semaphore(0);
+  private int totalTaskCount = 0;
+  private boolean acceptsTasks = true;
+
+  /** A single result from a background task. Encapsulates the task number to allow for sorting. */
+  private static class TaskResult<T> implements Comparable<TaskResult<T>> {
+    final int numTask;
+    final T result;
+    final Exception exception;
+
+    private TaskResult(int numTask, T result) {
+      this.numTask = numTask;
+      this.result = result;
+      this.exception = null;
+    }
+
+    private TaskResult(int numTask, Exception exception) {
+      this.numTask = numTask;
+      this.result = null;
+      this.exception = exception;
+    }
+
+    @Override
+    public int compareTo(TaskResult<T> other) {
+      return Integer.compare(numTask, other.numTask);
+    }
+  }
+
+  /** Enqueue a task on Android's THREAD_POOL_EXECUTOR. */
+  public void enqueueInBackground(Callable<T> task) {
+    hardAssert(acceptsTasks, "enqueueInBackground() called after awaitResults()");
+
+    int currentTask = ++totalTaskCount;
+    Executors.BACKGROUND_EXECUTOR.execute(
+        () -> {
+          try {
+            T result = task.call();
+            taskResults.add(new TaskResult<>(currentTask, result));
+          } catch (Exception e) {
+            taskResults.add(new TaskResult<>(currentTask, e));
+          }
+          completedTasks.release();
+        });
+  }
+
+  /** Executes a task inline, allowing for more efficient execution of small tasks. */
+  public void enqueueInline(Callable<T> callable) {
+    hardAssert(acceptsTasks, "enqueueInBackground() called after awaitResults()");
+
+    int currentTask = ++totalTaskCount;
+    try {
+      T result = callable.call();
+      taskResults.add(new TaskResult<>(currentTask, result));
+    } catch (Exception e) {
+      taskResults.add(new TaskResult<>(currentTask, e));
+    }
+    completedTasks.release();
+  }
+
+  /**
+   * Blocks on the execution of all tasks. Returns a list of results in the order that tasks where
+   * added.
+   */
+  public List<T> awaitResults() throws ExecutionException {
+    acceptsTasks = false;
+
+    List<T> allResults = new ArrayList<>(totalTaskCount);
+    try {
+      completedTasks.acquire(totalTaskCount);
+      while (!taskResults.isEmpty()) {
+        TaskResult<T> currentResult = taskResults.take();
+
+        if (currentResult.exception != null) {
+          throw new ExecutionException("Unhandled exception in task", currentResult.exception);
+        } else {
+          allResults.add(currentResult.result);
+        }
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```

Comment:
I worry slightly that this will return partial `allResults` if this method is interrupted. If the caller performs no further I/O or blocking operations, they may not see another exception and may proceed as if the contents of `allResults` are all the actual results. It seems risky to rely on something else to observe the interrupted status to conclude that the results are actually incorrect.

go/java-practices/interruptedexception recommends throwing an alternative exception (after re-interrupting). Should we just propagate InterruptedException here and let the caller figure it out? Alternatively, should we throw the appropriate FirebaseFirestoreException already here? It's going to be hard to describe what we were trying to do, so it seems like propagating is like the right thing.

Reply:
The page you linked to mostly deals with cancelled RPCs, in which case an Interrupt might be an expected event that needs to be treated explicitly. If we follow this advice, then awaitResults needs to re-throw some checked exception (either InterruptedException or FirebaseFirestoreException). This places all the burden on the callsite and again makes me question the usefulness of this class. 

If we do see an InterruptedException here, we can probably just assume that the client will no longer be in a useable state after (`acquire()` and take()` always run on our AsyncQueue). I'm trying to convince myself that throwing an AssertionFailure is the right thing to do here.